tests/testsuite/precise_pre_release.rs

```diff
@@ -90,30 +90,71 @@ fn update_pre_release_differ() {
             [package]
             name = "package"
             [dependencies]
-            my-dependency = "0.1.2"
+            my-dependency = "^0.1.2"
             "#,
         )
         .file("src/lib.rs", "")
         .build();
+    // "^0.1.2" is equivalent to ">=0.1.2, <0.2.0-0" according to `matches_prerelease_semantic`
 
+    // 0.1.2-pre.0 < 0.1.2 so it doesn't match
     p.cargo("update -p my-dependency --precise 0.1.2-pre.0 -Zunstable-options")
+        .masquerade_as_nightly_cargo(&["precise-pre-release"])
+        .with_status(101)
+        .with_stderr_data(str![[r#"
+[UPDATING] `dummy-registry` index
+[ERROR] failed to select a version for the requirement `my-dependency = "^0.1.2"`
+candidate versions found which didn't match: 0.1.2-pre.0
+location searched: `dummy-registry` index (which is replacing registry `crates-io`)
+required by package `package v0.0.0 ([ROOT]/foo)`
+if you are looking for the prerelease package it needs to be specified explicitly
+    my-dependency = { version = "0.1.2-pre.0" }
+perhaps a crate was updated and forgotten to be re-vendored?
+
+"#]])
+        .run();
+
+    // 0.2.0-0 is the upper bound we exclude, so it doesn't match
+    p.cargo("update -p my-dependency --precise 0.2.0-0 -Zunstable-options")
+        .masquerade_as_nightly_cargo(&["precise-pre-release"])
+        .with_status(101)
+        .with_stderr_data(str![[r#"
+[UPDATING] `dummy-registry` index
+[ERROR] failed to select a version for the requirement `my-dependency = "^0.1.2"`
+candidate versions found which didn't match: 0.2.0-0
+location searched: `dummy-registry` index (which is replacing registry `crates-io`)
+required by package `package v0.0.0 ([ROOT]/foo)`
+if you are looking for the prerelease package it needs to be specified explicitly
+    my-dependency = { version = "0.2.0-0" }
+perhaps a crate was updated and forgotten to be re-vendored?
+
+"#]])
+        .run();
+
+    // Test upgrade
+    // 0.1.3 is in the range, so it match
+    p.cargo("update -p my-dependency --precise 0.1.3 -Zunstable-options")
         .masquerade_as_nightly_cargo(&["precise-pre-release"])
         .with_stderr_data(str![[r#"
 [UPDATING] `dummy-registry` index
```

Comment:
Here should be some `upgrade` print, I remember which PR or issue mentioned this, didn't I

Reply:
My Bad. I shouldn't publish `0.1.3` in advance, the reason without `UPDATING` due to it alrealy is.